src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -708,5 +710,155 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+
+
+            // Software fallback
+            int tableIndex = (int)((crc ^ data) & 0xFF);
+            crc = s_crcTable[tableIndex] ^ (crc >> 8);
+
+            return crc;
+        }
+
+        /// <summary>
+        /// Calculates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
+        /// * uint32_t __crc32ch (uint32_t a, uint16_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, ushort data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint16_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(ushort)];
+            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
+
+            foreach (byte b in bytes)
+            {
+                int tableIndex = (int)((crc ^ b) & 0xFF);
+                crc = s_crcTable[tableIndex] ^ (crc >> 8);
+            }
+
+            return crc;
+        }
+
+        /// <summary>
+        /// Calculates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+        /// * uint32_t __crc32ch (uint32_t a, uint32_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, uint data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint32_t b)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(uint)];
+            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
+
+            foreach (byte b in bytes)
+            {
+                int tableIndex = (int)((crc ^ b) & 0xFF);
+                crc = s_crcTable[tableIndex] ^ (crc >> 8);
+            }
```

Comment:
These `stackalloc`s are inefficient and unnecessary.  Why not just operate directly on the value passed in?

And a known-size loop this small is probably best unrolled manually.  Something like [one of these](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgEllfCAAcAynwBugsDFzTajZi0XLV6gMIR8uwQBs+hqCbMWaMgDMTKQMjgwA3jQMMQwcIhgA2gC6DLgA+mBQYAAq2MCeDAC8DMIwAO5xCYmkAKxIyZaxDNGxxMHxymHZgaSOABSdGAxZYGhVXQAm2BjYAJStMVHUzc36utjCADzA4hgwAHwMu/u4xWmzYADW2O7uEGDHezCJuIIAXjAQXIMJc42LVYAVWEuGwvBYAHUoIJ9iDboIAOZlSb9WBcBgAWRg+Gg4kxYlwAAtbiwAOIwDDyGC8WDCMz9E7mObjaazOaWQHNLjQGDYMBEhiM57HBgiJ6nBYrVZLLkysVdWYFGBaSYwBDnfp/fr9UYMAB6xzmDAAZAw6AgAGKWjlymV6koZUZ5ZWJJWeVXq1KG3XZBgHI4ADlt0plAF8AqHVsQAOwjbJNWIR6iA9oTYaOHp9Ui/LqjcZDBhs+aA5bysUY/psWHOYRGPj7KCaXAAGVhGE8AFFhJNBJspeXmsXztXRBIAAow/CwwT1gTU+v8GDd3ubMq4XD9Yshu0MdHphjuDjDEr77G4iQE/gk9zkykAQSgUGw4gAIjNsNTaTB6TB+k7shdTwQ3LXceSgIUEjFc46EkaCtjSD4vh+IYOTFABqdCB3lMtBxiB0GBBMEIXvSZUX3I8MHGfphFQ4V9jmX1Hh9Jk5m3Y0fT1AMGGDRM8OHAMSkDPjVmTXdY3jMARLEqM00LTMwF6RxAlzYZ8wPbdS13QRK2rDBa0XRtmzbDAO2XHs+2EbD+I/EcRDEcRJ1UGc51sGBF1wczV2EddN23Tko2afdC0o84zxxPEr2JUkKQwR9nzfD8vz4H8GQA3J8mAkTml3AiiPBdRSPImlD2PajaO1VimINIVWPYmrqu43jdwEg4hOy2I8tBAqWCKtESso8q6KqvUWOeNiPzmDihS4oMQPlVr2tyv0SnykiyP6jFBqFCrlEYka/TGhj6s4v0mvmmVFp4jr8JWwjuvW4qtrKnbhvG6rDROmazrmgLywk0ZpJoMMgA==) (untested -- endianness might be backwards, but you should get the idea)

Reply:
> These stackallocs are inefficient and unnecessary. Why not just operate directly on the value passed in?
Can you explain me, why this is inefficient?

> And a known-size loop this small is probably best unrolled manually.
Why doesn't the JIT unroll loops this small automatically?

Reply:
> Can you explain me, why this is inefficient?

At present, the JIT always emits a stack buffer overrun check when you `stackalloc`.  Since your goal is only to write a single primitive value to the stack, there are multiple ways to do that without using `stackalloc`.  For example, you can create a span directly over the data argument like this:

```cs
ReadOnlySpan<uint> span = MemoryMarshal.CreateReadOnlySpan(ref data, 1);
```

Note the [codegen difference](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8YUCAAcAymIBuvMDFwBuGvSatOPfoIDCEfBN4AbMTKjzFKmmoDMTUg10MA3jQYeGHXjwDaAXQZcAH0wKDAAFWxgUwYAXgYuGAB3Lx8MX1IAViR/VWpPQoZ3T2J7bx4ncNtSXQAKCowGMLA0NMqAE2wMbABKEo83AqKPKQlsLgAeYABPDBgAPgZZ+dx4oJ6wAGtsY2MIMGW5mF9cXgAvGAgAMwb03ryBooBVLlxsa8EAdSheedfdrwAOZJDp1WDXBgAWRg+GgMyh2CguAAFrsWABxGAYNgwT6wLiKOorJS9NpdHq9fJPQrXaAwbBgFEMYnHZYMHxHVb9YYjIYjEbpBg9GIwEQdGAIdZ1e51OotBgAPWWvQYADIGHQEAAxbVUmkChUJEItKKi3wi0ziyWBZXy8IMBZLAAc+t5RQAvjZ3YViAB2ZrhfKFL3UJ5ldpNXTVWrBAByECkmy2AEF9gc7pUWm1GgwKX0nvyBbjsB0APJcYwzMYTSaNJa4cZcdYwuFQBFI1Ho6MM+Yl8uV6tN8F4vPdbBtchug0eOmwRnM1nzdmc1vwxHItHGFgp3AAIWOuDqjYmvR5AtcM6KQstYq4EqlCRlPF6coVyuAqo1Wt1bovhSNIJQnCM1TAtaIrXvG0lRZBUnQYV1gwFUMr39QMwCQjxQw9IA) with just that change.

> Why doesn't the JIT unroll loops this small automatically?

JIT has to balance the time it takes to identify and realize optimizations with the benefit they offer.  Loop unrolling is not always a win, and the JIT doesn't have the same amount of time, say, an optimizing C++ compiler does to evaluate that.  There is one place JIT will auto-unroll a loop because it's likely to always be of benefit, and that is loops over `Vector<T>.Count`, `Vector64<T>.Count`, `Vector128<T>.Count`, and `Vector256<T>.Count`.  You could take advantage of this fact to force an unroll of 4 with e.g. `for int i = 0; i < Vector128<int>.Count; i++`, but then in your code, you'd introduce bounds checks using `i` to index into the span.

The samples I put up yesterday eliminate both of these problems and remove the bounds checks for the CRC lookup table, so they would be significantly faster.

Reply:
Thank you for the clarification. Do you know why the JIT method ends with an `int3` instruction in both cases? Is it mandatory?

Reply:
The `call` that precedes the `int3` in both of those methods is the exception for the bounds check.  I believe JIT inserts the `int3` to indicate that the method does not actually return in that case.  Of course if you eliminate the bounds checks into the LUT, you also eliminate the code to throw on out of bounds, which is of course not possible when looking up in a 256-element array with a 1 byte index.